### PR TITLE
Add char matrices for Valencia

### PR DESCRIPTION
**Pull request by carmaza**
## Proposed changes

~Dependent on previous PRs. Please only review "Add characteristic matrices for Valencia" commit.~

~Depends on #1958.~ 

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: Doubles aren't allocated — they're placed on the stack and super cheap.

So I would strongly suggest favoring readability and stricter use of `const` by writing `const double normal_velocity = ...` here, `const double velocity_squared = ...` below, and so on elsewhere.

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: To make it more clear how this comment relates to the code, I suggest to rephrase this comment as "Set the eigenvector that has eigenvalue = ...", here and in similar comments below.

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp:79]: Possibly in the brief, but ideally also in the main body, say what this means: these vectors let you transform between the conserved and characteristic variables of the system.

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp]: For most users of the code, how this matrix is to be used will be more important than the precise expressions. I suggest that before giving the full mathematical expressions, you document the usage of "char = left * cons" and "cons = right * char", being clear about the ordering of the cons and char variables.

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp:106]: Add reference

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp:150]: All the same comments as for the right case.

File: tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp
Comment (fmahebert) [inline tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp]: I recommend testing with a Polytrope EoS too.

For this equation of state `kappa` goes to zero, so it's good to have a test that checks the choice of left/right vectors has no divisions by kappa. I did this wrong in the Newtonian case, and am currently working to fix it...

File: tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp
Comment (fmahebert) [inline tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp]: Please add test that "right * speeds * left = d(flux)/du".

(Or is this far too difficult in the relativistic case? I am concerned that the current test does not check that these matrices are giving the correct transform... only that the two matrices are inverses of each other)

**Review by fmahebert — CHANGES_REQUESTED** [review]

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp
Comment (carmaza) [reply src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: Back when I started working on this, I had these functions defined for a type different from `Matrix`, and the functions were templated in `DataType`, so allocations were important. Then I adopted the convention you implemented for the Newtonian case, which only works for `double`s, but decided to keep the code in case we decided to make it work for `DataVector`s as well. I can rework some terms a bit if you do not foresee any use for the matrices defined for `DataVector`s. What do you think?

File: tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp
Comment (carmaza) [reply tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp]: I've worked out some terms and I was planning on adding that test too. The calculation is lengthy, so please do not wait if you need this code in develop.

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp
Comment (fmahebert) [reply src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: I think your question is equivalent to this one: will we need to compute characteristic fields in a _pointwise_ way, i.e., with a characteristic transformation that is different at every grid point? I don't believe I've seen algorithms that expect this (aside from, e.g., the upwind flux for GH which we wouldn't want to use for hydro anyway). So my suspicion is that it is unlikely we would need this these matrices in terms of DataVectors... Is this consistent with your thoughts?

Edit to clarify: if you're aware of any cases that might require the char transformation to be computed pointwise, then I have no problem with allowing it in the code :)

File: tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp
Comment (fmahebert) [reply tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp]: Okay, good to hear. I do not plan to use this transformation just yet (and can always get it from this branch), so I'm fine with waiting to merge this PR until you finish working it out. But if you prefer to do this as two steps with two PRs, that's okay with me too.

File: tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp
Comment (teukolsky) [reply tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp]: Actually, in the relativistic case kappa = rho c_s^2 in this case (rho = rest mass). I think the difference is because of including rest mass in the energy density. The eigenvectors should be fine, the way I wrote them. It would be good to check numerically!

Comment (fmahebert) [reply tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp]: I haven't looked into this yet, but our `PolytropicFluid` equation of state sets `kappa_times_p_over_rho_squared` to 0 whether relativistic or not.

**Comment by carmaza** [comment]
With #1959 just merged, this PR should be easier to re-review. The fixups after @fmahebert 's review were a bit long so I split them into different commits in order to make them easier to read. Everything should be squashed after approving them.

This PR still depends on #1958 , which was already reviewed by @fmahebert but it still needs a second reviewer.

File: tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp
Comment (fmahebert) [inline tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp:140]: [nitpick] I suggest adding a blank line above this for readability

File: tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp
Comment (fmahebert) [inline tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp]: [nitpick] I suggest moving this down to line ~167 where it is first used.

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: [nitpick] I suggest moving this down to line ~ 136 above its first use

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: [nitpick] I suggest the simpler to read
```
const double d = ...;
const double sound_speed_over_d = sqrt(get(sound_speed_squared)) / d;
```

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp:202]: This complexity is surprising at first.. I suggest adding a short comment (above this line) similar to

"Here we use an overload of `orthonormal_oneform` that is only defined in 3D, so it must be called from within a `make_overloader`."

🙁

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: make -> set, for consistency with comments above

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: [nitpick] I suggest a more "active" name to clarify what this is doing... perhaps "set_degenerate_eigenvector"? ... up to you on this one.

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: ditto move down, to first use

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: ditto "make -> set"

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: ditto name of lambda

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp:352]: ditto comment on complexity

**Review by fmahebert — COMMENTED** [review]
Thanks @carmaza, I know my review was asking for some large changes in the test.

I'm happy with all your updates, so please go ahead and squash (also rebase to get the helpers that I had cherry-picked in a different PR) - up to you whether to do this now or after the other dependency is merged.

Below are a few minor nitpicks and comments regarding clarity; please squash these in directly while you squash/rebase.

**Comment by teukolsky** [comment]
I've checked the equations in the documentation - look fine.

**Comment by carmaza** [comment]
Squashed and rebased on develop.

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: `orthonormal_form` -> `orthonormal_oneform`

(here and in similar comment below)

**Review by fmahebert — COMMENTED** [review]
One small typo, otherwise LGTM! 👍 

Squash directly.

**Review by fmahebert — DISMISSED** [review]

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp
Comment (kidder) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp:159]: why are we not citing a published paper?

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp
Comment (kidder) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp:162]: can you either add commas between the elements or write this as a column vector with a `^T` to indicate transpose.  We find the current format hard to parse

**Review by kidder — COMMENTED** [review]
reviewed with @nilsdeppe

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp
Comment (carmaza) [reply src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp:162]: Added the commas to all left eigenvectors.

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp
Comment (carmaza) [reply src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp:159]: My understanding is that no other author has worked out this form of the eigenvectors in terms of an arbitrary normal orientation.

Comment (teukolsky) [reply src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp:159]: That's correct - these expressions are much simpler than those in the literature. They probably won't get published till I've finished the MHD ones, maybe a few months.

**Review by kidder — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Review by fmahebert — APPROVED** [review]

**Comment by nilsdeppe** [comment]
@fmahebert can you please let us know if you're happy with this when you get a chance? Thanks!

Edit: I was too slow, sorry!

**Pull request merged**
